File: ruby/lib/StdUriTemplate.rb
module StdUriTemplate

  # Public API
  public

  def self.expand(template, substitutions)
    expand_impl(template, substitutions)
  end

  # Private implementation
  private

  module Modifier
    NO_MOD = :no_mod
    PLUS = :plus
    DASH = :dash
    DOT = :dot
    SLASH = :slash
    SEMICOLON = :semicolon
    QUESTION_MARK = :question_mark
    AT = :at
  end

  def self.validate_literal(c, col)
    case c
    when '+', '#', '/', ';', '?', '&', ' ', '!', '=', '$', '|', '*', ':', '~', '-'
      raise ArgumentError, "Illegal character identified in the token at col:#{col}"
    end
  end

  def self.get_max_char(buffer, col)
    return -1 if buffer.nil?

    value = buffer.to_s

    if value.empty?
      -1
    else
      begin
        Integer(value)
      rescue ArgumentError
        raise ArgumentError, "Cannot parse max chars at col:#{col}"
      end
    end
  end

  def self.get_modifier(c, token, col)
    case c
    when '+'
      Modifier::PLUS
    when '#'
      Modifier::DASH
    when '.'
      Modifier::DOT
    when '/'
      Modifier::SLASH
    when ';'
      Modifier::SEMICOLON
    when '?'
      Modifier::QUESTION_MARK
    when '&'
      Modifier::AT
    else
      validate_literal(c, col)
      token << c
      Modifier::NO_MOD
    end
  end

  def self.expand_impl(str, substitutions)
    result = ''
    token = nil
    modifier = nil
    composite = false
    max_char_buffer = nil
    first_token = true

    str.chars.each_with_index do |character, i|
      case character
      when '{'
        token = ''
        first_token = true
      when '}'
        if token
          expanded = expand_token(modifier, token, composite, get_max_char(max_char_buffer, i), first_token, substitutions, result, i)
          first_token = false if expanded && first_token
          token = nil
          modifier = nil
          composite = false
          max_char_buffer = nil
        else
          raise ArgumentError, "Failed to expand token, invalid at col:#{i}"
        end
      when ','
        if token
          expanded = expand_token(modifier, token, composite, get_max_char(max_char_buffer, i), first_token, substitutions, result, i)
          first_token = false if expanded && first_token
          token = ''
          composite = false
          max_char_buffer = nil
        end
      else
        if token
          if modifier.nil?
            modifier = get_modifier(character, token, i)
          elsif max_char_buffer
            if character =~ /\d/
              max_char_buffer << character
            else
              raise ArgumentError, "Illegal character identified in the token at col:#{i}"
            end
          else
            if character == ':'
              max_char_buffer = ''
            elsif character == '*'
              composite = true
            else
              validate_literal(character, i)
              token << character
            end
          end
        else
          result << character
        end
      end
    end

    if token.nil?
      result
    else
      raise ArgumentError, 'Unterminated token'
    end
  end

  def self.add_prefix(mod, result)
    case mod
    when Modifier::DASH
      result << '#'
    when Modifier::DOT
      result << '.'
    when Modifier::SLASH
      result << '/'
    when Modifier::SEMICOLON
      result << ';'
    when Modifier::QUESTION_MARK
      result << '?'
    when Modifier::AT
      result << '&'
    end
  end

  def self.add_separator(mod, result)
    case mod
    when Modifier::DOT
      result << '.'
    when Modifier::SLASH
      result << '/'
    when Modifier::SEMICOLON
      result << ';'
    when Modifier::QUESTION_MARK, Modifier::AT
      result << '&'
    else
      result << ','
    end
  end

  def self.add_value(mod, token, value, result, max_char)
    case mod
    when Modifier::PLUS, Modifier::DASH
      add_expanded_value(value, result, max_char, false)
    when Modifier::QUESTION_MARK, Modifier::AT
      result << token + '='
      add_expanded_value(value, result, max_char, true)
    when Modifier::SEMICOLON
      result << token
      result << '=' unless value.empty?
      add_expanded_value(value, result, max_char, true)
    when Modifier::DOT, Modifier::SLASH, Modifier::NO_MOD
      add_expanded_value(value, result, max_char, true)
    end
  end

  def self.add_value_element(mod, token, value, result, max_char)
    case mod
    when Modifier::PLUS, Modifier::DASH
      add_expanded_value(value, result, max_char, false)
    when Modifier::QUESTION_MARK, Modifier::AT, Modifier::SEMICOLON, Modifier::DOT, Modifier::SLASH, Modifier::NO_MOD
      add_expanded_value(value, result, max_char, true)
    end
  end

  def self.add_expanded_value(value, result, max_char, replace_reserved)
    max = (max_char != -1) ? [max_char, value.length].min : value.length
    reserved_buffer = nil

    max.times do |i|
      character = value[i]

      if character == '%' && !replace_reserved
        reserved_buffer = ''
      end

      if reserved_buffer
        reserved_buffer << character

        if reserved_buffer.length == 3
          is_encoded = false
          begin
            encoded = URI.decode_www_form_component(reserved_buffer)
            is_encoded = !(encoded == reserved_buffer)
          rescue StandardError
          end

          if is_encoded
            result << reserved_buffer
          else
            result << "%25"
            result << reserved_buffer[1..-1] unless replace_reserved
          end

          reserved_buffer = nil
        end
      else
        if character == ' '
          result << "%20"
        elsif character == '%'
          result << "%25"
        else
          if replace_reserved
            result << URI.encode_www_form_component(character)
          else
            result << character
          end
        end
      end
    end

    if reserved_buffer
      result << "%25"
      if replace_reserved
        result << URI.encode_www_form_component(reserved_buffer[1..-1])
      else
        result << reserved_buffer[1..-1]
      end
    end
  end

  def self.list?(value)
    value.kind_of?(Array) || value.kind_of?(Enumerable)
  end

  def self.map?(value)
    value.kind_of?(Hash)
  end

  module SubstitutionType
    STRING = :string
    LIST = :list
    MAP = :map
  end

  def self.get_substitution_type(value, col)
    if (value.nil? || value.kind_of?(String))
      SubstitutionType::STRING
    elsif map?(value)
      SubstitutionType::MAP
    elsif list?(value)
      SubstitutionType::LIST
    else
      raise ArgumentError, "Illegal class passed as substitution, found #{value.class} at col:#{col}"
    end
  end

  def self.empty?(subst_type, value)
    case subst_type
    when SubstitutionType::STRING
      value.nil?
    when SubstitutionType::LIST
      value.respond_to?(:empty?) ? value.empty? : true
    when SubstitutionType::MAP
      value.respond_to?(:empty?) ? value.empty? : true
    else
      true
    end
  end

  def self.expand_token(modifier, token, composite, max_char, first_token, substitutions, result, col)
    raise ArgumentError, "Found an empty token at col:#{col}" if token.empty?

    value = substitutions[token]
    value = value.to_s if [Integer, Float].any? { |type| value.is_a?(type) }

    subst_type = get_substitution_type(value, col)
    return false if empty?(subst_type, value)

    if first_token
      add_prefix(modifier, result)
    else
      add_separator(modifier, result)
    end

    case subst_type
    when SubstitutionType::STRING
      add_string_value(modifier, token, value.to_s, result, max_char)
    when SubstitutionType::LIST
      add_list_value(modifier, token, value, result, max_char, composite)
    when SubstitutionType::MAP
      add_map_value(modifier, token, value, result, max_char, composite)
    end

    true
  end

  def self.add_string_value(modifier, token, value, result, max_char)
    add_value(modifier, token, value, result, max_char)
  end

  def self.add_list_value(modifier, token, value, result, max_char, composite)
    first = true
    value.each do |v|
      if first
        add_value(modifier, token, v.to_s, result, max_char)
        first = false
      else
        if composite
          add_separator(modifier, result)
          add_value(modifier, token, v.to_s, result, max_char)
        else
          result << ','
          add_value_element(modifier, token, v.to_s, result, max_char)
        end
      end
    end
    !first
  end

  def self.add_map_value(modifier, token, value, result, max_char, composite)
    first = true
    raise ArgumentError, 'Value trimming is not allowed on Maps' if max_char != -1

    value.each do |key, val|
      if composite
        add_separator(modifier, result) unless first
        add_value_element(modifier, token, key.to_s, result, max_char)
        result << '='
      else
        if first
          add_value(modifier, token, key.to_s, result, max_char)
        else
          result << ','
          add_value_element(modifier, token, key.to_s, result, max_char)
        end
        result << ','
      end
      add_value_element(modifier, token, val.to_s, result, max_char)
      first = false
    end
    !first
  end
end

File: csharp/StdUriTemplate.cs
#nullable disable
namespace stduritemplate;

using System;
using System.Collections;
using System.Collections.Generic;
using System.Text;

public class StdUriTemplate
{
    // Public API
    public static string Expand(string template, Dictionary<string, object> substitutions)
    {
        return ExpandImpl(template, substitutions);
    }

    // Private implementation
    private enum Modifier
    {
        NO_MOD,
        PLUS,
        DASH,
        DOT,
        SLASH,
        SEMICOLON,
        QUESTION_MARK,
        AT
    }

    private static void ValidateLiteral(char c, int col)
    {
        switch (c)
        {
            case '+':
            case '#':
            case '/':
            case ';':
            case '?':
            case '&':
            case ' ':
            case '!':
            case '=':
            case '$':
            case '|':
            case '*':
            case ':':
            case '~':
            case '-':
                throw new ArgumentException($"Illegal character identified in the token at col:{col}");
            default:
                break;
        }
    }

    private static int GetMaxChar(StringBuilder buffer, int col)
    {
        if (buffer == null || buffer.Length < 1)
        {
            return -1;
        }

        try
        {
            return int.Parse(buffer.ToString());
        }
        catch (FormatException)
        {
            throw new ArgumentException($"Cannot parse max chars at col:{col}");
        }
    }

    private static Modifier GetModifier(char c, StringBuilder token, int col)
    {
        switch (c)
        {
            case '+': return Modifier.PLUS;
            case '#': return Modifier.DASH;
            case '.': return Modifier.DOT;
            case '/': return Modifier.SLASH;
            case ';': return Modifier.SEMICOLON;
            case '?': return Modifier.QUESTION_MARK;
            case '&': return Modifier.AT;
            default:
                ValidateLiteral(c, col);
                token.Append(c);
                return Modifier.NO_MOD;
        }
    }

    private static string ExpandImpl(string str, Dictionary<string, object> substitutions)
    {
        StringBuilder result = new StringBuilder(str.Length * 2);

        StringBuilder token = null;

        Modifier? modifier = null;
        bool composite = false;
        StringBuilder maxCharBuffer = null;
        bool firstToken = true;

        for (int i = 0; i < str.Length; i++)
        {
            char character = str[i];
            switch (character)
            {
                case '{':
                    token = new StringBuilder();
                    firstToken = true;
                    break;
                case '}':
                    if (token != null)
                    {
                        bool expanded = ExpandToken(modifier, token.ToString(), composite, GetMaxChar(maxCharBuffer, i), firstToken, substitutions, result, i);
                        if (expanded && firstToken)
                        {
                            firstToken = false;
                        }
                        token = null;
                        modifier = null;
                        composite = false;
                        maxCharBuffer = null;
                    }
                    else
                    {
                        throw new ArgumentException($"Failed to expand token, invalid at col:{i}");
                    }
                    break;
                case ',':
                    if (token != null)
                    {
                        bool expanded = ExpandToken(modifier, token.ToString(), composite, GetMaxChar(maxCharBuffer, i), firstToken, substitutions, result, i);
                        if (expanded && firstToken)
                        {
                            firstToken = false;
                        }
                        token = new StringBuilder(token.Length * 2);
                        composite = false;
                        maxCharBuffer = null;
                        break;
                    }
                    // Intentional fall-through for commas outside the {}
                    goto default;
                default:
                    if (token != null)
                    {
                        if (modifier == null)
                        {
                            modifier = GetModifier(character, token, i);
                        }
                        else if (maxCharBuffer != null)
                        {
                            if (char.IsDigit(character))
                            {
                                maxCharBuffer.Append(character);
                            }
                            else
                            {
                                throw new ArgumentException($"Illegal character identified in the token at col:{i}");
                            }
                        }
                        else
                        {
                            if (character == ':')
                            {
                                maxCharBuffer = new StringBuilder(3);
                            }
                            else if (character == '*')
                            {
                                composite = true;
                            }
                            else
                            {
                                ValidateLiteral(character, i);
                                token.Append(character);
                            }
                        }
                    }
                    else
                    {
                        result.Append(character);
                    }
                    break;
            }
        }

        if (token == null)
        {
            return result.ToString();
        }
        else
        {
            throw new ArgumentException("Unterminated token");
        }
    }

    private static void AddPrefix(Modifier? mod, StringBuilder result)
    {
        switch (mod)
        {
            case Modifier.DASH:
                result.Append('#');
                break;
            case Modifier.DOT:
                result.Append('.');
                break;
            case Modifier.SLASH:
                result.Append('/');
                break;
            case Modifier.SEMICOLON:
                result.Append(';');
                break;
            case Modifier.QUESTION_MARK:
                result.Append('?');
                break;
            case Modifier.AT:
                result.Append('&');
                break;
            default:
                return;
        }
    }

    private static void AddSeparator(Modifier? mod, StringBuilder result)
    {
        switch (mod)
        {
            case Modifier.DOT:
                result.Append('.');
                break;
            case Modifier.SLASH:
                result.Append('/');
                break;
            case Modifier.SEMICOLON:
                result.Append(';');
                break;
            case Modifier.QUESTION_MARK:
            case Modifier.AT:
                result.Append('&');
                break;
            default:
                result.Append(',');
                break;
        }
    }

    private static void AddValue(Modifier? mod, string token, string value, StringBuilder result, int maxChar)
    {
        switch (mod)
        {
            case Modifier.PLUS:
            case Modifier.DASH:
                AddExpandedValue(value, result, maxChar, false);
                break;
            case Modifier.QUESTION_MARK:
            case Modifier.AT:
                result.Append(token + '=');
                AddExpandedValue(value, result, maxChar, true);
                break;
            case Modifier.SEMICOLON:
                result.Append(token);
                if (!string.IsNullOrEmpty(value))
                {
                    result.Append('=');
                }
                AddExpandedValue(value, result, maxChar, true);
                break;
            case Modifier.DOT:
            case Modifier.SLASH:
            case Modifier.NO_MOD:
                AddExpandedValue(value, result, maxChar, true);
                break;
        }
    }

    private static void AddValueElement(Modifier? mod, string token, string value, StringBuilder result, int maxChar)
    {
        switch (mod)
        {
            case Modifier.PLUS:
            case Modifier.DASH:
                AddExpandedValue(value, result, maxChar, false);
                break;
            case Modifier.QUESTION_MARK:
            case Modifier.AT:
            case Modifier.SEMICOLON:
            case Modifier.DOT:
            case Modifier.SLASH:
            case Modifier.NO_MOD:
                AddExpandedValue(value, result, maxChar, true);
                break;
        }
    }

    private static void AddExpandedValue(string value, StringBuilder result, int maxChar, bool replaceReserved)
    {
        int max = (maxChar != -1) ? Math.Min(maxChar, value.Length) : value.Length;
        result.EnsureCapacity(max * 2); // hint to SB
        StringBuilder reservedBuffer = null;

        for (int i = 0; i < max; i++)
        {
            char character = value[i];

            if (character == '%' && !replaceReserved)
            {
                reservedBuffer = new StringBuilder(3);
            }

            if (reservedBuffer != null)
            {
                reservedBuffer.Append(character);

                if (reservedBuffer.Length == 3)
                {
                    bool isEncoded = false;
                    try
                    {
                        var original = reservedBuffer.ToString();
                        isEncoded = !original.Equals(Uri.UnescapeDataString(original));
                    }
                    catch (Exception)
                    {
                        // ignore
                    }

                    if (isEncoded)
                    {
                        result.Append(reservedBuffer);
                    }
                    else
                    {
                        result.Append("%25");
                        // only if !replaceReserved
                        result.Append(reservedBuffer.ToString(1, 2));
                    }
                    reservedBuffer = null;
                }
            }
            else
            {
                if (character == ' ')
                {
                    result.Append("%20");
                }
                else if (character == '%')
                {
                    result.Append("%25");
                }
                else
                {
                    if (replaceReserved)
                    {
                        result.Append(Uri.EscapeDataString(character.ToString()));
                    }
                    else
                    {
                        result.Append(character);
                    }
                }
            }
        }

        if (reservedBuffer != null)
        {
            result.Append("%25");
            if (replaceReserved)
            {
                result.Append(Uri.EscapeDataString(reservedBuffer.ToString(1, reservedBuffer.Length - 1)));
            }
            else
            {
                result.Append(reservedBuffer.ToString(1, reservedBuffer.Length - 1));
            }
        }
    }

    private static bool IsList(object value)
    {
        return value is IList;
    }

    private static bool IsDictionary(object value)
    {
        return value is IDictionary;
    }

    private enum SubstitutionType
    {
        STRING,
        LIST,
        DICTIONARY
    }

    private static SubstitutionType GetSubstitutionType(object value, int col)
    {
        if (value is string || value == null)
        {
            return SubstitutionType.STRING;
        }
        else if (IsList(value))
        {
            return SubstitutionType.LIST;
        }
        else if (IsDictionary(value))
        {
            return SubstitutionType.DICTIONARY;
        }
        else
        {
            throw new ArgumentException($"Illegal class passed as substitution, found {value.GetType()} at col:{col}");
        }
    }

    private static bool IsEmpty(SubstitutionType substType, object value)
    {
        if (value == null)
        {
            return true;
        }
        else
        {
            switch (substType)
            {
                case SubstitutionType.STRING:
                    return value == null;
                case SubstitutionType.LIST:
                    return ((IList)value).Count == 0;
                case SubstitutionType.DICTIONARY:
                    return ((IDictionary)value).Count == 0;
                default:
                    return true;
            }
        }
    }

    // returns true if expansion happened
    private static bool ExpandToken(
            Modifier? modifier,
            string token,
            bool composite,
            int maxChar,
            bool firstToken,
            Dictionary<string, object> substitutions,
            StringBuilder result,
            int col)
    {
        if (string.IsNullOrEmpty(token))
        {
            throw new ArgumentException($"Found an empty token at col:{col}");
        }

        object value;
        substitutions.TryGetValue(token, out value);
        if (value is bool ||
                value is int ||
                value is long ||
                value is float ||
                value is double)
        {
            value = value.ToString().ToLower();
        } else if (value is DateTime dt)
        {
            value = dt.ToUniversalTime().ToString("yyyy-MM-dd'T'HH:mm:ssK");
        }

        SubstitutionType substType = GetSubstitutionType(value, col);
        if (IsEmpty(substType, value))
        {
            return false;
        }

        if (firstToken)
        {
            AddPrefix(modifier, result);
        }
        else
        {
            AddSeparator(modifier, result);
        }

        switch (substType)
        {
            case SubstitutionType.STRING:
                AddStringValue(modifier, token, (string)value, result, maxChar);
                break;
            case SubstitutionType.LIST:
                AddListValue(modifier, token, (IList)value, result, maxChar, composite);
                break;
            case SubstitutionType.DICTIONARY:
                AddDictionaryValue(modifier, token, ((IDictionary)value), result, maxChar, composite);
                break;
        }

        return true;
    }

    private static bool AddStringValue(Modifier? modifier, string token, string value, StringBuilder result, int maxChar)
    {
        AddValue(modifier, token, value, result, maxChar);
        return true;
    }

    private static bool AddListValue(Modifier? modifier, string token, IList value, StringBuilder result, int maxChar, bool composite)
    {
        bool first = true;
        foreach (string v in value)
        {
            if (first)
            {
                AddValue(modifier, token, v, result, maxChar);
                first = false;
            }
            else
            {
                if (composite)
                {
                    AddSeparator(modifier, result);
                    AddValue(modifier, token, v, result, maxChar);
                }
                else
                {
                    result.Append(',');
                    AddValueElement(modifier, token, v, result, maxChar);
                }
            }
        }
        return !first;
    }

    private static bool AddDictionaryValue(Modifier? modifier, string token, IDictionary value, StringBuilder result, int maxChar, bool composite)
    {
        bool first = true;
        if (maxChar != -1)
        {
            throw new ArgumentException("Value trimming is not allowed on Dictionaries");
        }
        foreach (DictionaryEntry v in value)
        {
            if (composite)
            {
                if (!first)
                {
                    AddSeparator(modifier, result);
                }
                AddValueElement(modifier, token, (string)v.Key, result, maxChar);
                result.Append('=');
            }
            else
            {
                if (first)
                {
                    AddValue(modifier, token, (string)v.Key, result, maxChar);
                }
                else
                {
                    result.Append(',');
                    AddValueElement(modifier, token, (string)v.Key, result, maxChar);
                }
                result.Append(',');
            }
            AddValueElement(modifier, token, (string)v.Value, result, maxChar);
            first = false;
        }
        return !first;
    }
}
